### PR TITLE
Optimize python scripts

### PR DESCRIPTION
**Pull request by noobOriented**
# Goal
Improve the performance of `data_acquisition_shadow.py`, make it ready for larger inputs.

# Results
benchmark scripts:

```bash
$ python data_acquisition_shadow.py derandomized 5 tests/snapshots/generated_observables_20.txt
```

Before (7a133e1): 48.07s
After: 0.07s  **~700x** 🚀 

# Changes
1. Replace loops with numpy array operation, which utilizes [BLAS](https://www.netlib.org/blas/)
1. Sparse operations: only perform computation on data that can change the result
   1. Pop the observables that already have enough matches [code](https://github.com/noobOriented/predicting-quantum-properties/pull/2/files#r1886402655)
   1. For each qubit, ignore the observable that does not contribute the difference to loss.  [code](https://github.com/noobOriented/predicting-quantum-properties/pull/2/files#r1886398917)
1. Do the broadcasting carefully [code](https://github.com/noobOriented/predicting-quantum-properties/pull/2/files#r1886401695)
   1. if `a`, `b` have different dimesions: `np.matmul(a, b)` is much faster than `np.sum(a * b)`. The latter produces intermediate product array.
   1. Do the expensive operation first: `exp(a_i + log(b_ij))` -> `exp(a_i) * b_ij`, the input size of `exp` is smaller

### CODE REVIEW

File: tests/snapshots/data_acquisition_shadow_d_m10_system_size5.txt [hunk 1]
```diff
@@ -65,8 +65,8 @@ Z Z Z X X
 Z X X Y Y
 Z Y Y X X
 Z Z X X Z
-Y Y X X X
```
Comment (noobOriented) [inline tests/snapshots/data_acquisition_shadow_d_m10_system_size5.txt:68]: XXX: unsure if the order matters here

File: data_acquisition_shadow.py [hunk 2]
```diff
@@ -84,114 +91,98 @@ def derandomized_classical_shadow(
...
+
+        num_of_measurements[finished_qubits] += 1
+        keep_indices, = np.nonzero(num_of_measurements < weight * num_of_measurements_per_observable)
+        if len(keep_indices) == 0:
+            return
+
+        num_of_measurements = num_of_measurements[keep_indices]
+        observables = observables[keep_indices]
+        observable_counts = observable_counts[keep_indices]
+        weight = weight[keep_indices]
```
Comment (noobOriented) [inline data_acquisition_shadow.py]: Changes 2.i

File: data_acquisition_shadow.py [hunk 3]
```diff
@@ -84,114 +91,98 @@ def derandomized_classical_shadow(
...
+        observable_counts = np.count_nonzero(observables, axis=1)
+
+    matches = np.zeros([len(observables)])  # shape (N, )
+    measurement: list[int] = []
+
+    # find best op for each qubit
+    for pos in range(observables.shape[1]):
+        # 1. When observables[i, pos] == 0, its contribution to cost is independent to op thus can be ignored
+        # 2. Once matches[i] becomes -inf, logits never change no matter what op you choose
+        indices, = np.nonzero((observables[:, pos] != 0) & (matches != -np.inf))  # shape (M, )
```
Comment (noobOriented) [inline data_acquisition_shadow.py:123]: changes 2.ii

File: data_acquisition_shadow.py [hunk 4]
```diff
@@ -84,114 +91,98 @@ def derandomized_classical_shadow(
...
+    matches_needed: npt.NDArray,  # shape (3, N)
+    weights: npt.NDArray,  # shape (N,)
+    *,
+    eta: float = 0.9,
+) -> npt.NDArray:
+    nu = 1 - np.exp(-eta / 2)
+    logits = (
+        -n_measurements * eta / 2
+        + np.log(1 - nu / np.pow(3, matches_needed))
+    ) / weights  # shape (3, N)
```
Comment (noobOriented) [inline data_acquisition_shadow.py]: changes 4

File: data_acquisition_shadow.py [hunk 5]
```diff
@@ -50,148 +50,105 @@ def randomized_classical_shadow(
...
+):
+    if observable_counts is None:
+        observable_counts = np.count_nonzero(observables, axis=1)
+
+    matches = np.zeros([len(observables)])  # shape (N, )
+    measurement: list[int] = []
+
+    # find best op for each qubit
+    for pos in range(observables.shape[1]):
+        # 1. When observables[i, pos] == 0, its contribution to cost is independent to op thus can be ignored
```
Comment (noobOriented) [inline data_acquisition_shadow.py:121]: Changes 2.ii

File: data_acquisition_shadow.py [hunk 6]
```diff
@@ -50,148 +50,105 @@ def randomized_classical_shadow(
...
 
-    return measurement_procedure
+def cost_func(
+    n_measurements: npt.NDArray,  # shape (N,)
+    matches_needed: npt.NDArray,  # shape (3, N)
+    *,
+    eta: float = 0.9,
+) -> npt.NDArray:  # shape (3,)
+    nu = 1 - np.exp(-eta / 2)
+    return (1 - nu / (3 ** matches_needed)) @ np.exp(-n_measurements * (eta / 2))
```
Comment (noobOriented) [inline data_acquisition_shadow.py:151]: Changes 3.i & 3.ii

File: data_acquisition_shadow.py [hunk 7]
```diff
@@ -50,148 +50,105 @@ def randomized_classical_shadow(
...
+        yield measurement
+
+        num_of_measurements[finished_qubits] += 1
+        keep_indices, = np.nonzero(num_of_measurements < num_of_measurements_per_observable)
+        if len(keep_indices) == 0:
+            return
+
+        num_of_measurements = num_of_measurements[keep_indices]
+        observables = observables[keep_indices]
+        observable_counts = observable_counts[keep_indices]
```
Comment (noobOriented) [inline data_acquisition_shadow.py:105]: Changes 2.i
